File: 30-substring-with-concatenation-of-all-words/substring-with-concatenation-of-all-words.java
import java.util.*;

class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        List<Integer> result = new ArrayList<>();
        if (s == null || words == null || words.length == 0 || s.length() == 0) {
            return result;
        }

        int wordLength = words[0].length();
        int wordCount = words.length;
        int totalWordsLength = wordLength * wordCount;
        
        // Create a hashmap for word frequencies in the list
        Map<String, Integer> wordMap = new HashMap<>();
        for (String word : words) {
            wordMap.put(word, wordMap.getOrDefault(word, 0) + 1);
        }

        // Iterate over possible starting points
        for (int i = 0; i < wordLength; i++) {
            Map<String, Integer> currentCount = new HashMap<>();
            int left = i;
            int count = 0;

            // Slide through the string `s`
            for (int j = i; j + wordLength <= s.length(); j += wordLength) {
                String word = s.substring(j, j + wordLength);
                
                if (wordMap.containsKey(word)) {
                    currentCount.put(word, currentCount.getOrDefault(word, 0) + 1);
                    count++;
                    
                    // If a word is seen more times than it appears in `words`, shift left pointer
                    while (currentCount.get(word) > wordMap.get(word)) {
                        String leftWord = s.substring(left, left + wordLength);
                        currentCount.put(leftWord, currentCount.get(leftWord) - 1);
                        left += wordLength;
                        count--;
                    }

                    // If all words match, add the starting index to the result
                    if (count == wordCount) {
                        result.add(left);
                    }
                } else {
                    // Reset the window if the word is not in `words`
                    currentCount.clear();
                    count = 0;
                    left = j + wordLength;
                }
            }
        }

        return result;
    }
}

File: 30-substring-with-concatenation-of-all-words/substring-with-concatenation-of-all-words.cpp
#include <vector>
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
        vector<int> result;
        if (s.empty() || words.empty()) return result;

        int wordLength = words[0].length();
        int totalWordsLen = wordLength * words.size();
        int n = s.length();

        // Create a frequency map for the words
        unordered_map<string, int> wordCount;
        for (const string& word : words) {
            wordCount[word]++;
        }

        // Loop through possible offsets (0 to wordLength - 1)
        for (int offset = 0; offset < wordLength; offset++) {
            unordered_map<string, int> currentCount;
            int start = offset, matchedWords = 0;

            for (int end = offset; end + wordLength <= n; end += wordLength) {
                // Get the current word
                string word = s.substr(end, wordLength);

                // If it's a valid word, update the current count
                if (wordCount.find(word) != wordCount.end()) {
                    currentCount[word]++;
                    
                    // Check if the count is valid
                    if (currentCount[word] <= wordCount[word]) {
                        matchedWords++;
                    } else {
                        // Remove extra occurrences
                        while (currentCount[word] > wordCount[word]) {
                            string removedWord = s.substr(start, wordLength);
                            currentCount[removedWord]--;
                            if (currentCount[removedWord] < wordCount[removedWord]) {
                                matchedWords--;
                            }
                            start += wordLength;
                        }
                    }

                    // If all words match, add the start index
                    if (matchedWords == words.size()) {
                        result.push_back(start);
                    }
                } else {
                    // Reset the window if the word is invalid
                    currentCount.clear();
                    matchedWords = 0;
                    start = end + wordLength;
                }
            }
        }

        return result;
    }
};
